File: JavaScript/array/deque-dual.js
/**
 * Author: Axel Ariel Saravia
 */

import ArrayStack from "./stack.js";

/**
DualDeque T :: {
    front: Stack T
    back: Stack T
}
*/

const DualArrayDeque = Object.freeze({
    /**
    create T :: () -> DualDeque T */
    create() {
        return Object.seal({
            front: ArrayStack.create(),
            back: ArrayStack.create()
        });
    },
    /**
    balance T :: (DualDeque T) -> DualDeque T */
    balance(deque) {
        const n = DualArrayDeque.size(deque);
        const mid = Math.floor(n / 2);
        if (
            3 * deque.front.length < deque.back.length
            || 3  * deque.back.length < deque.front.length
        ) {
            const f = ArrayStack.create();
            const b = ArrayStack.create();
            for (let i = 0; i < mid; i += 1) {
                ArrayStack.add(
                    f,
                    i,
                    DualArrayDeque.get(deque, mid - i - 1)
                );
            }
            for (let i = 0; i < n - mid; i += 1) {
                ArrayStack.add(
                    b,
                    i,
                    DualArrayDeque.get(deque, mid + i)
                );
            }
            deque.front = f;
            deque.back = b;
        }
        return deque;
    },
    /**
    size T :: (DualDeque T) -> number[uint] */
    size(deque) {
        return deque.front.length + deque.back.length
    },
    /**
    get T :: (DualDeque T, number [uint]) -> maybe T */
    get(deque, i) {
        return (
            i < deque.front.length
            ? ArrayStack.get(front, deque.front.length - i - 1)
            : this.back.get(i - this.front.size())
        );
    },
    /**
    add T :: (DualDeque T, number [uint], T) -> DualDeque T */
    add(deque, i, x) {
        if (i < deque.front.length) {
            ArrayStack.add(
                deque.front,
                deque.front.length - i,
                x
            );
        } else {
            ArrayStack.add(
                deque.back,
                i - deque.front.length,
                x
            );
        }
        return DualArrayDeque.balance(deque);
    },
    /**
    remove T :: (DualDeque T, number [uint]) -> T */
    remove(deque, i) {
        const x = (
            i < deque.front.length
            ? ArrayStack.remove(deque.front, deque.front.length - i - 1)
            : ArrayStack.remove(deque.back, i - deque.front.length)
        );
        DualArrayDeque.balance(deque);
        return x;
    }
});

export default DualArrayDeque;

File: C/array/deque-dual.c
/**
 * Author: Axel Ariel Saravia
 */

/*
    IMPORTANT

    the T type is the signed type

    dualDeque struct is {
        stack front;
        stack back;
    }
*/


# include "deque-dual.h"

dualDeque dual_deque_create() {
    stack front = stack_create();
    stack back = stack_create();
    return (
        !front.content || !back.content
        ? (dualDeque){0}
        : (dualDeque){
            .front = front,
            .back = back,
        }
    );
}

size_t dual_deque_size(dualDeque const ddq) {
    return ddq.front.length + ddq.back.length;
}

T dual_deque_get(dualDeque const ddq, size_t const i) {
    size_t const length = ddq.front.length;
    size_t idx = 0;
    if (i < length) {
        idx = length - i - 1;
        return stack_get(ddq.front, idx);
    } else {
        idx = i - length;
        return stack_get(ddq.back, idx);
    }
}

T dual_deque_set(dualDeque ddq[static 1], size_t const i, T const x) {
    size_t const length = ddq->front.length;
    size_t idx = 0;
    if (i < length) {
        idx = length - i - 1;
        return stack_set(&(ddq->front), idx, x);
    } else {
        idx = i - length;
        return stack_set(&(ddq->back), idx, x);
    }
}

dualDeque dual_deque_add(dualDeque ddq, size_t const i, T const x) {
    size_t const length = ddq.front.length;
    size_t idx = 0;
    if (i < length) {
        idx = length - i;
        ddq.front = stack_add(ddq.front, idx, x);
    } else {
        idx =  i - length;
        ddq.back = stack_add(ddq.back, idx, x);
    }
    return dual_deque_balance(ddq);
}

dualDeque dual_deque_remove(dualDeque ddq, size_t const i) {
    size_t const length = ddq.front.length;
    size_t idx = 0;
    if (i < length) {
        idx = length - i - 1;
        ddq.front = stack_remove(ddq.front, idx);
    } else {
        idx =  i - length;
        ddq.back = stack_remove(ddq.back, idx);
    }
    return dual_deque_balance(ddq);
}

dualDeque dual_deque_balance(dualDeque ddq) {
    size_t const flength = ddq.front.length;
    size_t const blength = ddq.back.length;
    size_t mid = (flength + blength) / 2;
    if (3 * flength < blength || 3 * blength < flength) {
        stack front = stack_create();
        stack back = stack_create();
        for (size_t i = 0; i < mid; i += 1) {
            size_t idx = mid - i - 1;
            T x = dual_deque_get(ddq, idx);
            front = stack_add(front, i, x);
        }
        for (size_t i = 0; i < flength + blength - mid; i += 1) {
            size_t idx = mid + i;
            T x = dual_deque_get(ddq, idx);
            back = stack_add(back, i, x);
        }
        ddq.front = front;
        ddq.back = back;
    }
    return ddq;
}

void dual_deque_free(dualDeque ddq[static 1]) {
    stack_free(&(ddq->front));
    stack_free(&(ddq->back));
    *ddq = (dualDeque){0};
}